File: 2025-11-November-GFG-POTD/November 21 - Shortest Path Using Atmost One Curved Edge/Solution.py
import heapq

class Solution:
    INF = 10**18

    def dijkstra(self, src, V, adj):
        dist = [self.INF] * V
        dist[src] = 0

        pq = [(0, src)]  # (distance, node)

        while pq:
            d, u = heapq.heappop(pq)

            if d != dist[u]:
                continue

            for v, w in adj[u]:
                if d + w < dist[v]:
                    dist[v] = d + w
                    heapq.heappush(pq, (dist[v], v))

        return dist

    def shortestPath(self, V, a, b, edges):
        adj = [[] for _ in range(V)]

        # Add straight edges
        for x, y, w1, w2 in edges:
            adj[x].append((y, w1))
            adj[y].append((x, w1))

        distA = self.dijkstra(a, V, adj)
        distB = self.dijkstra(b, V, adj)

        ans = distA[b]  # no curved edge

        # Try using curved edges once
        for x, y, w1, w2 in edges:
            # use curved x → y
            if distA[x] < self.INF and distB[y] < self.INF:
                ans = min(ans, distA[x] + w2 + distB[y])

            # use curved y → x
            if distA[y] < self.INF and distB[x] < self.INF:
                ans = min(ans, distA[y] + w2 + distB[x])

        return -1 if ans >= self.INF else ans

File: 2025-11-November-GFG-POTD/November 21 - Shortest Path Using Atmost One Curved Edge/Solution.java
class Solution {
     static class Pair {
        int v, w;
        Pair(int v, int w) { this.v = v; this.w = w; }
    }

    static final int INF = Integer.MAX_VALUE;

    private int[] dijkstra(int src, int V, ArrayList<ArrayList<Pair>> adj) {
        int[] dist = new int[V];
        Arrays.fill(dist, INF);

        PriorityQueue<Pair> pq = new PriorityQueue<>((a,b) -> a.w - b.w);
        pq.add(new Pair(src, 0));
        dist[src] = 0;

        while (!pq.isEmpty()) {
            Pair cur = pq.poll();
            int u = cur.v, d = cur.w;

            if (d != dist[u]) continue;  // stale entry

            for (Pair p : adj.get(u)) {
                int v = p.v, w = p.w;

                if (d + w < dist[v]) {
                    dist[v] = d + w;
                    pq.add(new Pair(v, dist[v]));
                }
            }
        }
        return dist;
    }

    public int shortestPath(int V, int a, int b, int[][] edges) {

        // Graph with ONLY straight edges
        ArrayList<ArrayList<Pair>> adj = new ArrayList<>();
        for (int i = 0; i < V; i++) adj.add(new ArrayList<>());

        // Build adjacency for straight edges
        for (int[] e : edges) {
            int x = e[0], y = e[1], w1 = e[2];
            adj.get(x).add(new Pair(y, w1));
            adj.get(y).add(new Pair(x, w1));
        }

        // Dijkstra from a and b
        int[] distA = dijkstra(a, V, adj);
        int[] distB = dijkstra(b, V, adj);

        long ans = distA[b];  // case: no curved edge

        // Try using each curved edge ONCE
        for (int[] e : edges) {
            int x = e[0], y = e[1], w2 = e[3];

            if (distA[x] != INF && distB[y] != INF)
                ans = Math.min(ans, (long) distA[x] + w2 + distB[y]);

            if (distA[y] != INF && distB[x] != INF)
                ans = Math.min(ans, (long) distA[y] + w2 + distB[x]);
        }

        return ans >= INF ? -1 : (int) ans;
    }
}
